packages/hyperion-autologging/src/ALSurfaceVisibilityPublisher.ts

```diff
@@ -127,21 +135,123 @@ export function publish(options: InitOptions): void {
+                    const childElement = node as Element;
+                    const surfaceDataList = observedRoots.get(childElement);
+
+                    // Check if this was a child we were observing
+                    if (surfaceDataList && surfaceDataList.includes(surfaceData)) {
+                      console.log('Child removed', childElement);
+
+                      // Stop observing the removed child
+                      observer.unobserve(childElement);
+                      observedRoots.delete(childElement, surfaceData);
+                      surfaceDataRoots.delete(surfaceData, childElement);
+
+                      // Start observing the parent element again
+                      observer.observe(element);
+                      observedRoots.set(element, surfaceData);
+                      surfaceDataRoots.set(surfaceData, element);
+                    }
+                  }
+                }
+              }
+            }
+
+            // Then handle added nodes
+            let hasViableChild = false;
+            if (mutation.addedNodes.length > 0) {
+              // Check if any of the added nodes are elements we can observe
+              for (let i = 0; i < mutation.addedNodes.length; i++) {
+                const node = mutation.addedNodes[i];
+                if (node.nodeType === Node.ELEMENT_NODE) {
+                  const childElement = node as Element;
+                  console.log('Child found', childElement);
+
+                  if (!ALSurfaceUtils.isSurfaceWrapper(childElement)) {
+                    // Check if we're currently observing the parent element
+                    const surfaceDataList = observedRoots.get(element);
+                    if (surfaceDataList && surfaceDataList.includes(surfaceData)) {
+                      // Found a viable child, switch observation to it
+                      observer.unobserve(element);
+                      observer.observe(childElement);
+
+                      // Update tracking maps
+                      observedRoots.delete(element, surfaceData);
+                      observedRoots.set(childElement, surfaceData);
+                      surfaceDataRoots.delete(surfaceData, element);
+                      surfaceDataRoots.set(surfaceData, childElement);
+                      hasViableChild = true;
+                      break;
+                    }
+                  }
+                }
+              }
+            }
+            // If we are tracking removal,  we need to keep the observer around,  which could be potentially expensive
+            if (hasViableChild && !enableDynamicChildTrackingForRemoval) {
+              // Disconnect the mutation observer as we no longer need it
+              mutationObserver.disconnect();
+              mutationObservers.delete(element);
+            }
+          }
+        }
+      });
+
+      // Start observing for child additions
+      mutationObserver.observe(element, { childList: true });
```

Comment:
The logic of mutation observer is fixed for all elements. so we should not create new one every time. We just need one (no need for `mutationObservers`) and just decide what to `observe` or `unobserve` an element's mutations. 

I think if the subtree added one after another, you might get only of mutation event with a long nodeList, correct?
In that case, you need find the highest node in the tree and use that as the sub-tree root.
So, you need to map nodes to surfaceData, then for all nodes under a surface, find the node that contains all others ones in the list, then add that one as the root.

Reply:
Let me think on how I can make it a single one,  right now there are variables scoped into the mutation observer callback,  such as the root element,  `surfaceData` and the current `observer`.

Reply:
Oh, I didn't realize it is not turning on subtree monitoring. That's actually better.
In that case, you can simplify the logic and just say, whenever the child node is added, repeat the logic of a) loop and recurse to find a suitable node, if not, b) add another mutation observer?

Basically, recursively only look at the immediate children until you find something.